resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -535,18 +553,32 @@ private InetSocketAddress nextNameServerAddress() {
             InetSocketAddress nameServerAddr, DnsQuestion question,
             Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> promise) {
 
-        return query0(checkNotNull(nameServerAddr, "nameServerAddr"),
-                      checkNotNull(question, "question"),
-                      checkNotNull(promise, "promise"));
+        return query0(nameServerAddr, question, null, promise);
+    }
+
+    /**
+     * Sends a DNS query with the specified question with additional records using the specified name server list.
+     */
+    public Future<AddressedEnvelope<DnsResponse, InetSocketAddress>> query(
+            InetSocketAddress nameServerAddr, DnsQuestion question,
+            List<DnsRecord> additional,
+            Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> promise) {
+
+        return query0(nameServerAddr, question, additional, promise);
     }
 
     private Future<AddressedEnvelope<DnsResponse, InetSocketAddress>> query0(
             InetSocketAddress nameServerAddr, DnsQuestion question,
+            List<DnsRecord> additional,
             Promise<AddressedEnvelope<? extends DnsResponse, InetSocketAddress>> promise) {
 
-        final Promise<AddressedEnvelope<DnsResponse, InetSocketAddress>> castPromise = cast(promise);
+        final Promise<AddressedEnvelope<DnsResponse, InetSocketAddress>> castPromise = cast(
+                checkNotNull(promise, "promise"));
         try {
-            new DnsQueryContext(this, nameServerAddr, question, castPromise).query();
+            new DnsQueryContext(this,
+                    checkNotNull(nameServerAddr, "nameServerAddr"),
+                    checkNotNull(question, "question"),
+                    additional, castPromise).query();
```

Comment:
Actually, I think all those checks would be better directly in `DnsQueryContext ` constructor. This way, you could get rid of your checks in `addRecord`.